### PR TITLE
Poll types

### PR DESCRIPTION
**Pull request by DevAndyLee**
Refactored Voting JavaScript classes to use Knockout instead of jQuery DOM manipulation, and implemented unit tests.

### CODE REVIEW

File: VotingApplication/VotingApplication.Web.Api/Scripts/BasicVote.js [hunk 1]
```diff
@@ -152,47 +83,62 @@
...
+        };
+
+        self.initialise = function (pollData) {
+            self.pollOptions.initialise(pollData);
+        };
 
-        self.addOption = function () {
-            //Don't submit without an entry in the name field
-            if ($("#newName").val() === "") {
+        // TODO: Extract chart code from viewModel class - ideally
```
Comment (tpkelly) [inline VotingApplication/VotingApplication.Web.Api/Scripts/BasicVote.js]: We should file this as a JIRA issue and remove the TODO comment

Comment (DevAndyLee) [inline VotingApplication/VotingApplication.Web.Api/Scripts/BasicVote.js]: It's what I'm planning to do next under VA-183 - Remove view code from view model
https://int.scottlogic.co.uk/jira/browse/VA-183

File: VotingApplication/VotingApplication.Web.Api/Scripts/PointsVote.js [hunk 2]
```diff
@@ -222,73 +108,82 @@
...
+            self.maxPerVote(pollData.MaxPerVote);
+            self.maxPoints(pollData.MaxPoints);
 
-        self.addOption = function () {
-            //Don't submit without an entry in the name field
-            if ($("#newName").val() === "") {
+            resetVote();
+        }
+
+        // TODO: Extract chart code from viewModel class - ideally
```
Comment (tpkelly) [inline VotingApplication/VotingApplication.Web.Api/Scripts/PointsVote.js]: Same again

File: VotingApplication/VotingApplication.Web.Api/Scripts/RankedVote.js [hunk 3]
```diff
@@ -170,14 +144,112 @@
...
+
+                    // Cancel the sort operation and update the Knockout
+                    // arrays, letting Knockout re-arrange the DOM
+                    $(".sortable").sortable("cancel");
+                    selectPickedOptions(votes);
+                }
+            });
+        };
+
+        // TODO: Extract chart code from viewModel class - ideally
```
Comment (tpkelly) [inline VotingApplication/VotingApplication.Web.Api/Scripts/RankedVote.js]: Same again

File: VotingApplication/VotingApplication.Web.Api/Scripts/BasicVote.js (hunk 1 above)
Comment (tpkelly) [inline VotingApplication/VotingApplication.Web.Api/Scripts/BasicVote.js]: In that case, can we remove the TODO?

Comment (DevAndyLee) [inline VotingApplication/VotingApplication.Web.Api/Scripts/BasicVote.js]: Any particular reason? I moved the code to the end and added the comment to make it clear that it wasn't, and shouldn't be, part of the viewModel code. Personally I think it's valid to keep that comment there until the problem is resolved.

**Pull request merged**